File: ios/SsdidWallet/Domain/OpenId4Vci/TokenClient.swift
import Foundation

/// Response from a token endpoint exchange.
struct TokenResponse: CustomStringConvertible {
    let accessToken: String
    let tokenType: String
    let cNonce: String?
    let cNonceExpiresIn: Int?

    var description: String {
        "TokenResponse(accessToken=REDACTED, tokenType=\(tokenType), cNonce=\(cNonce != nil ? "present" : "nil"))"
    }
}

/// Handles OAuth 2.0 token exchange for OpenID4VCI flows.
final class TokenClient: @unchecked Sendable {

    private let session: URLSession

    init(session: URLSession = .shared) {
        self.session = session
    }

    /// Exchanges a pre-authorized code for an access token.
    func exchangePreAuthorizedCode(
        tokenEndpoint: String,
        preAuthorizedCode: String,
        txCode: String? = nil
    ) async throws -> TokenResponse {
        var params: [(String, String)] = [
            ("grant_type", "urn:ietf:params:oauth:grant-type:pre-authorized_code"),
            ("pre-authorized_code", preAuthorizedCode)
        ]
        if let txCode = txCode {
            params.append(("tx_code", txCode))
        }
        return try await postTokenRequest(tokenEndpoint: tokenEndpoint, params: params)
    }

    /// Exchanges an authorization code for an access token.
    func exchangeAuthorizationCode(
        tokenEndpoint: String,
        code: String,
        codeVerifier: String,
        redirectUri: String
    ) async throws -> TokenResponse {
        let params: [(String, String)] = [
            ("grant_type", "authorization_code"),
            ("code", code),
            ("code_verifier", codeVerifier),
            ("redirect_uri", redirectUri)
        ]
        return try await postTokenRequest(tokenEndpoint: tokenEndpoint, params: params)
    }

    // MARK: - Private

    private func postTokenRequest(
        tokenEndpoint: String,
        params: [(String, String)]
    ) async throws -> TokenResponse {
        guard let url = URL(string: tokenEndpoint) else {
            throw OpenId4VciError.transportError("Invalid token endpoint URL: \(tokenEndpoint)")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")

        let body = params.map { key, value in
            let encodedKey = key.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? key
            let encodedValue = value.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? value
            return "\(encodedKey)=\(encodedValue)"
        }.joined(separator: "&")
        request.httpBody = body.data(using: .utf8)

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse,
              (200..<300).contains(httpResponse.statusCode) else {
            let code = (response as? HTTPURLResponse)?.statusCode ?? -1
            throw OpenId4VciError.tokenError("Token request failed: HTTP \(code)")
        }

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw OpenId4VciError.tokenError("Invalid token response JSON")
        }

        guard let accessToken = json["access_token"] as? String else {
            throw OpenId4VciError.tokenError("Missing access_token")
        }

        return TokenResponse(
            accessToken: accessToken,
            tokenType: json["token_type"] as? String ?? "Bearer",
            cNonce: json["c_nonce"] as? String,
            cNonceExpiresIn: json["c_nonce_expires_in"] as? Int
        )
    }
}

File: android/app/src/main/java/my/ssdid/wallet/domain/oid4vci/TokenClient.kt
package my.ssdid.wallet.domain.oid4vci

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request

data class TokenResponse(
    val accessToken: String,
    val tokenType: String,
    val cNonce: String?,
    val cNonceExpiresIn: Int?
) {
    override fun toString() = "TokenResponse(accessToken=<redacted>, tokenType=$tokenType, cNonce=$cNonce, cNonceExpiresIn=$cNonceExpiresIn)"
}

class TokenClient(private val client: OkHttpClient) {

    fun exchangePreAuthorizedCode(
        tokenEndpoint: String,
        preAuthorizedCode: String,
        txCode: String? = null
    ): Result<TokenResponse> = runCatching {
        val formBuilder = FormBody.Builder()
            .add("grant_type", "urn:ietf:params:oauth:grant-type:pre-authorized_code")
            .add("pre-authorized_code", preAuthorizedCode)
        if (txCode != null) formBuilder.add("tx_code", txCode)

        postTokenRequest(tokenEndpoint, formBuilder.build())
    }

    fun exchangeAuthorizationCode(
        tokenEndpoint: String,
        code: String,
        codeVerifier: String,
        redirectUri: String
    ): Result<TokenResponse> = runCatching {
        val form = FormBody.Builder()
            .add("grant_type", "authorization_code")
            .add("code", code)
            .add("code_verifier", codeVerifier)
            .add("redirect_uri", redirectUri)
            .build()

        postTokenRequest(tokenEndpoint, form)
    }

    private fun postTokenRequest(tokenEndpoint: String, formBody: FormBody): TokenResponse {
        val request = Request.Builder()
            .url(tokenEndpoint)
            .post(formBody)
            .build()

        return client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                throw RuntimeException("Token request failed: HTTP ${response.code}")
            }
            val body = response.body?.string()
                ?: throw RuntimeException("Empty token response")
            val json = Json.parseToJsonElement(body).jsonObject

            TokenResponse(
                accessToken = json["access_token"]?.jsonPrimitive?.content
                    ?: throw IllegalStateException("Missing access_token"),
                tokenType = json["token_type"]?.jsonPrimitive?.content ?: "Bearer",
                cNonce = json["c_nonce"]?.jsonPrimitive?.contentOrNull,
                cNonceExpiresIn = json["c_nonce_expires_in"]?.jsonPrimitive?.intOrNull
            )
        }
    }
}
